pwnlib/fmtstr.py

```diff
@@ -102,7 +108,497 @@ def send_payload(payload):
+        >>> pp(pwnlib.fmtstr.find_min_hamming_in_range(4, 0x1234, 0x10000, 0x0))
+        (3, '0x10000', '0xff00ffff')
+    """
+    steps = []
+    for _ in xrange(maxbytes):
+        steps += [(lower, upper, target)]
+        lower = lower >> 8
+        upper = upper >> 8
+        target = target >> 8
+
+    # the initial state
+    prev = {
+        (False,False): (0, 0, 0),
+        (False,True): None if upper == lower else (0, lower, 0),
+        (True,False): None if upper == lower else (0, lower, 0),
+        (True,True): None if upper <= lower + 1 else (0, lower + 1, 0)
+    }
+    for step in reversed(steps):
+        prev = {
+            (carry, strict): find_min_hamming_in_range_step(prev, step, carry, strict )
+            for carry in [False, True]
+            for strict in [False, True]
+        }
+    return prev[(False,False)]
+#
+# what we don't do:
+#  - create new atoms that cannot be created by merging existing atoms
+#  - optimize based on masks
+def merge_atoms_overlapping(atoms, sz, szmax, numbwritten, overflows):
+    if not szmax:
+        szmax = max(SPECIFIER.keys())
+
+    assert 1 <= overflows, "must allow at least one overflow"
+    assert sz <= szmax, "sz must be smaller or equal to szmax"
+
+    maxwritten = numbwritten + (1 << (8 * sz)) * overflows
+    done = [False for _ in atoms]
+
+    def possible_writes(idx, numbwritten):
+        candidate = AtomWrite(atoms[idx].start, 0, 0)
+        best = 0
+        for nextidx in xrange(idx, len(atoms)):
+            if candidate.size > szmax or done[nextidx] or candidate.end != atoms[nextidx].start:
+                break
+
+            candidate = candidate.union(atoms[nextidx])
+            if candidate.size not in SPECIFIER: continue
+
+            approxed = candidate
+            score = candidate.size
+            if approxed.size > sz:
+                score, v, m = find_min_hamming_in_range(approxed.size, numbwritten, maxwritten, approxed.integer)
+                approxed = candidate.replace(integer=v, mask=m)
+
+            if score > best:
+                best = score
+                yield list(xrange(idx, nextidx+1)), approxed
+
+    numbwritten_at = [numbwritten for _ in atoms]
+    out = []
+    for idx in xrange(len(atoms)):
+        if done[idx]: continue
+        numbwritten_here = numbwritten_at[idx]
+        idxrange, candidate = list(possible_writes(idx, numbwritten_here))[-1]
+        numbwritten_here += candidate.compute_padding(numbwritten_here)
+        offset = 0
+        for i in idxrange:
+            shift = atoms[i].size
+            if not ((~candidate[offset:offset+shift].mask) & atoms[i].mask):
+                done[i] = True
+            else:
+                numbwritten_at[i] = max(numbwritten_at[i], numbwritten_here)
+            offset += shift
+        out += [candidate]
+    return out
+
+def sort_atoms(atoms, numbwritten):
+    # find dependencies
+    order = { atom: i for i,atom in enumerate(atoms) }
+
+    def conflicts():
```

Comment:
It looks like you're using nested functions here.  Please consider using a class with properties and methods instead.